[addons][vis][gui] improve visualization support

AlwinEsch opened the pull request:
## Description

---------------------
**Commit 1:**
### [addons][vis] change AddonInstance_Visualization to use pointers
This is intended to maintain backward compatibility when making changes.

Before, all the data was together which made adding things difficult and always required a complete version upgrade.

---------------------
**Commit 2:**
### [addons][vis] fix address set of is_dirty function

Before was this address not set and bring mistakes to Kodi.

---------------------
**Commit 3:**
### [addons][vis][gui] use IsDirty addon call again 
Before was the IsDirty never called as it was alone in a unused Function.

This shifts it to the "Process" call and if "IsDirty" returns true the "MarkDirtyRegion" is called. Most addons do not use the IsDirty and therefore is always true as default, so no big changes occur.

However, if needed now, the IsDirty can also be used on the addon.

---------------------
**Commit 4:**
### [addons][vis] add function to transfer presets and increase version to 2.0.2

This is unfortunately necessary because of a small hen and egg problem. When using Angle to Windows, the GL is only supported at the startup call.

To ProjectM e.g. the presets can only be captured after the `new projectM` call, which in turn requires GL.
Hereby it is possible to transmit this during the "Start" call from the addon.

The new clear function is only used within the headers to avoid flooding on multiple calls.

<!--- Provide a general summary of your change in the Pull Request title above -->
<!--- Describe your change in detail here -->

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here -->

## How Has This Been Tested?
<!--- Please describe in detail how you tested your change -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc -->

## Screenshots (if appropriate):

## Types of change
<!--- What type of change does your code introduce? Put an `x` in all the boxes that apply like this: [X] -->
- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [ ] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:
<!--- Go over all the following points, and put an `X` in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [ ] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

AlwinEsch committed:
[addons][vis] change AddonInstance_Visualization to use pointers

This is intended to maintain backward compatibility when making changes.

Before, all the data was together which made adding things difficult and always required a complete version upgrade.

AlwinEsch committed:
[addons][vis][gui] use IsDirty addon call again

Before was the IsDirty never called as it was alone in a unused Function.

This shifts it to the "Process" call and if "IsDirty" returns true the "MarkDirtyRegion" is called. Most addons do not use the IsDirty and therefore is always true as default, so no big changes occur.

However, if needed now, the IsDirty can also be used on the addon.

Rechi commented on xbmc/addons/Visualization.cpp:
```diff
@@ -152,4 +153,16 @@ void CVisualization::transfer_preset(void* kodiInstance, const char* preset)
 
+void CVisualization::clear_presets(void* kodiInstance)
+{
+  CVisualization *addon = static_cast<CVisualization*>(kodiInstance);
```
> Left-align `*` and `&` to the base type they modify.

https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#52-pointer-and-reference-types

AlwinEsch committed:
[addons][vis] add function to transfer presets and increase version to 2.0.2

This is unfortunately necessary because of a small hen and egg problem. When using Angle to Windows, the GL is only supported at the startup call.

To ProjectM e.g. the presets can only be captured after the `new projectM` call, which in turn requires GL.
Hereby it is possible to transmit this during the "Start" call from the addon.

The new clear function is only used within the headers to avoid flooding on multiple calls.

AlwinEsch committed:
[addons][vis] fix address set of is_dirty function

Before was this address not set and bring mistakes to Kodi.

AlwinEsch commented on xbmc/addons/kodi-addon-dev-kit/include/kodi/addon-instance/Visualization.h:
```diff
@@ -656,6 +656,7 @@ namespace addon
       m_instanceData->toAddon->start = ADDON_Start;
       m_instanceData->toAddon->stop = ADDON_Stop;
       m_instanceData->toAddon->audio_data = ADDON_AudioData;
+      m_instanceData->toAddon->is_dirty = ADDON_IsDirty;
```
Found here also a big bug, the address was never set therefore the function was never used.

I just noticed now. You should not have the debug view on the screen when testing :joy: (this then make his "IsDirty" always itself :smirk: and therefore seen only now).

AchimTuran commented on xbmc/addons/Visualization.cpp:
```diff
@@ -143,4 +153,16 @@ void CVisualization::transfer_preset(void* kodiInstance, const char* preset)
 
+void CVisualization::clear_presets(void* kodiInstance)
+{
+  CVisualization* addon = static_cast<CVisualization*>(kodiInstance);
```
pls use reinterpret_cast for void pointers

AlwinEsch replied:
About this I'm unsure and better to leave on `static_cast`, seen always that `reinterpet`should only be used if nothing else works.

Know the following:
https://stackoverflow.com/questions/332030/when-should-static-cast-dynamic-cast-const-cast-and-reinterpret-cast-be-used
https://stackoverflow.com/questions/573294/when-to-use-reinterpret-cast
https://stackoverflow.com/questions/4476182/which-cast-to-use-static-cast-or-reinterpret-cast

AchimTuran replied:
You're right! You should leave it as static_cast. I was wrong, I thought it is good practice to use reinterpret_cast for void pointer casts.
